add e2etest for aggregated api endpoint

wwwnay opened the pull request:
**What type of PR is this?**

<!--
Add one of the following kinds:

/kind feature


-->

**What this PR does / why we need it**:
Add e2e test case for aggregated-api-endpoint
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```
NONE
```

RainbowMango commented:
Wow! Big thanks!!! @wwwnay 

/assign @XiShanYongYe-Chang

XiShanYongYe-Chang commented:
@wwwnay thanks!

Maybe you can signoff refer to https://github.com/karmada-io/karmada/pull/1400/checks?check_run_id=5331140890

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,157 @@
+
+var _ = ginkgo.Describe("aggregatedapi testing", func() {
+	resourceNames := []string{}
+	clustername1 := "member1"
```
Recommend use Camelback naming, just like: `clusterName2`.

Same as other verbs naming.

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,157 @@
+	apisPath1 := "/apis/cluster.karmada.io/v1alpha1/clusters/" + clustername1 + "/proxy/apis"
+	apisPath2 := "/apis/cluster.karmada.io/v1alpha1/clusters/" + clustername2 + "/proxy/apis"
+	proxyclustrole := &rbacv1.ClusterRole{
+		TypeMeta:   metav1.TypeMeta{},
```
Unassigned fields can be omitted.

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,157 @@
+		}},
+	}
+
+	member1config := "/root/.kube/members.config"
+	karmadaconfig := "/root/.kube/karmada.config"
```
We can use `karmadaClient` to connect `karmada-apiserver`; and use func `GetClusterClient(clusterName string)` to get client of the member cluster.

wwwnay replied:
@XiShanYongYe-Chang how can i get the karmada-apiserver Host with karmadaClient
or is there other way i can get the Host

XiShanYongYe-Chang replied:
We can use:
https://github.com/karmada-io/karmada/blob/0365d04c5dae741cb2517237dda28cc03ead987d/test/e2e/suite_test.go#L68
just like:
https://github.com/karmada-io/karmada/blob/0365d04c5dae741cb2517237dda28cc03ead987d/test/e2e/failover_test.go#L63

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,157 @@
+		ObjectMeta: metav1.ObjectMeta{Name: saName},
+		Secrets:    []corev1.ObjectReference{},
+	}
+	//servive account ClusterRoleBinding
```
Usually, we add a space to the beginning of the comment, just as:
```
// servive account ClusterRoleBinding
```

XiShanYongYe-Chang reviewed (commented):
Good job 👍 

When you've updated, please cc me, I will review going, thanks!

wwwnay commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,177 @@
+	})
+}
+
+// Delete clusterrole.
+func Removeclusterrole(client kubernetes.Interface, name string) {
```
Thanks for the help! 
I will modify   according to your suggestion and the format

XiShanYongYe-Chang commented:
Hi @wwwnay, have you occur any problems, maybe we can discuss.

wwwnay commented:
@XiShanYongYe-Chang I have update ,please review

XiShanYongYe-Chang commented:
Hi @wwwnay, have you tested it on your local site? I test it failed on my side.

wwwnay commented:
> Hi @wwwnay, have you tested it on your local site? I test it failed on my side.

yes, i  run test on my local fine ,what's the error report 
`Image: "swr.ap-southeast-1.myhuaweicloud.com/karmada/karmada-interpreter-webhook-example:latest" with ID "sha256:c90abff56a62e9544e2724cdc4c564d2e7777ad050e99eb1339ca1eb004d8070" found to be already present on all nodes.
Switched to context "karmada-host".
deployment.apps/karmada-interpreter-webhook-example created
service/karmada-interpreter-webhook-example created
serviceaccount/karmada-interpreter-webhook-example created
wait the karmada-interpreter-webhook-example ready...
pod/karmada-interpreter-webhook-example-667548b7b5-dd5w9 condition met
Switched to context "karmada-apiserver".
resourceinterpreterwebhookconfiguration.config.karmada.io/examples created
customresourcedefinition.apiextensions.k8s.io/workloads.workload.example.io created
Switched to context "member1".
customresourcedefinition.apiextensions.k8s.io/workloads.workload.example.io created
Switched to context "member2".
customresourcedefinition.apiextensions.k8s.io/workloads.workload.example.io created
Switched to context "member3".
customresourcedefinition.apiextensions.k8s.io/workloads.workload.example.io created
Running Suite: E2E Suite
========================
Random Seed: 1646718390
Will run 5 of 5 specs

I0308 13:46:34.997614 2606507 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:107
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:113
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:120
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
resource name may not be empty
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:127
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom)
STEP: Creating member clusterrolebinding(tom)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:135
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
Ran 5 of 5 Specs in 2.556 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS`

wwwnay commented:
> Hi @wwwnay, have you tested it on your local site? I test it failed on my side.

yes,the test have passed on my side,this is the result.
`I0308 13:46:34.997614 2606507 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:107
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:113
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:120
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
resource name may not be empty
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:127
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom)
STEP: Creating member clusterrolebinding(tom)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:135
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found
•
Ran 5 of 5 Specs in 2.556 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS`

XiShanYongYe-Chang commented:
> Passed | 0 Failed | 0 Pending | 0 Skipped PASS

No cases appear to be executed.

You can directly use commend `hack/run-e2e.sh` under karmada path.

In order to focus on someone target, you can update like this:

```diff
--- a/hack/run-e2e.sh
+++ b/hack/run-e2e.sh
@@ -38,7 +38,7 @@ export KUBECONFIG=${KARMADA_APISERVER_KUBECONFIG}
 export PULL_BASED_CLUSTERS=${PULL_BASED_CLUSTERS}
 
 set +e
-ginkgo -v -race -failFast ./test/e2e/
+ginkgo -v -race -failFast -focus="aggregatedapi testing" ./test/e2e/
 TESTING_RESULT=$?
```

Error info on my side:

```
------------------------------
aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test 
  Serviceaccount  access in the member1 cluster
  /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:107
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
STEP: Remove sa(tom)
STEP: Remove sa(tom)
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
clusterroles.rbac.authorization.k8s.io "tom" not found
clusterrolebindings.rbac.authorization.k8s.io "tom" not found

• Failure [0.150 seconds]
aggregatedapi testing
/root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:13
  Serviceaccount Unified authentication Test
  /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:105
    Serviceaccount Unified authentication Test
    /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:106
      Serviceaccount  access in the member1 cluster [It]
      /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:107

      Expected
          <int>: 403
      to equal
          <int>: 200

      /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:110
------------------------------
```

wwwnay commented:
> > Passed | 0 Failed | 0 Pending | 0 Skipped PASS
> 
> No cases appear to be executed.
> 
> You can directly use commend `hack/run-e2e.sh` under karmada path.
> 
> In order to focus on someone target, you can update like this:
> 
> ```diff
> --- a/hack/run-e2e.sh
> +++ b/hack/run-e2e.sh
> @@ -38,7 +38,7 @@ export KUBECONFIG=${KARMADA_APISERVER_KUBECONFIG}
>  export PULL_BASED_CLUSTERS=${PULL_BASED_CLUSTERS}
>  
>  set +e
> -ginkgo -v -race -failFast ./test/e2e/
> +ginkgo -v -race -failFast -focus="aggregatedapi testing" ./test/e2e/
>  TESTING_RESULT=$?
> ```
> 
> Error info on my side:
> 
> ```
> ------------------------------
> aggregatedapi testing Serviceaccount Unified authentication Test Serviceaccount Unified authentication Test 
>   Serviceaccount  access in the member1 cluster
>   /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:107
> STEP: Creating serviceaccount(tom) in membercluster (member1)
> STEP: Creating serviceaccount(tom) in karmadaCluster
> STEP: Creating ClusterRole(cluster-proxy-clusterrole)
> STEP: Creating ClusterRole(cluster-proxy-clusterrolebinding)
> STEP: Remove sa(tom)
> STEP: Remove sa(tom)
> STEP: Remove ClusterRole(cluster-proxy-clusterrole)
> STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
> clusterroles.rbac.authorization.k8s.io "tom" not found
> clusterrolebindings.rbac.authorization.k8s.io "tom" not found
> 
> • Failure [0.150 seconds]
> aggregatedapi testing
> /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:13
>   Serviceaccount Unified authentication Test
>   /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:105
>     Serviceaccount Unified authentication Test
>     /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:106
>       Serviceaccount  access in the member1 cluster [It]
>       /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:107
> 
>       Expected
>           <int>: 403
>       to equal
>           <int>: 200
> 
>       /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:110
> ------------------------------
> ```

thanks, I run about 20 times and failed 2 times  ,i will check what happened when it failed

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,190 @@
+func GetSecretInfo(client kubernetes.Interface, karmadaClient karmada.Interface, saname string) (token string, url string) {
+	var secretName string
+	sclist := []string{}
+	listSecret, err := client.CoreV1().Secrets("default").List(context.TODO(), metav1.ListOptions{})
```
We can get target secret through `ServiceAccount.secrets` field.  And maybe the target secret will take some time to generate, so we need to wait it use `gomega.Eventually`.

wwwnay replied:
sorry, i didn't see the comment before  I repush  ,i will fix this  today

XiShanYongYe-Chang replied:
It's okay.

wwwnay commented:
/retest

wwwnay commented:
> 

yes ,i run this successful on my local site.
`
Running Suite: E2E Suite
========================
Random Seed: 1647254966
Will run 5 of 5 specs

I0314 18:49:31.251656 2317773 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:104
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:112
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:121
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:142
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:151
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
Ran 5 of 5 Specs in 4.783 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

`

wwwnay commented:
> 

yes,i run it successful on my local site.

wwwnay commented:
`
Running Suite: E2E Suite
========================
Random Seed: 1647254966
Will run 5 of 5 specs

I0314 18:49:31.251656 2317773 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:104
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:112
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:121
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:142
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:151
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
Ran 5 of 5 Specs in 4.783 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

`

wwwnay commented:
```
Running Suite: E2E Suite
========================
Random Seed: 1647254966
Will run 5 of 5 specs

I0314 18:49:31.251656 2317773 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:104
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:112
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:121
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for secret (tom) present in cluster
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:142
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:151
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)
•
Ran 5 of 5 Specs in 4.783 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,161 @@
+			gomega.Expect(code).Should(gomega.Equal(403))
+		})
+	})
+})
```
Add a blank line below the line.

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,227 @@
+	ginkgo.By(fmt.Sprintf("wait for secret (%s) present in cluster ", saname), func() {
+		saresult, err := client.CoreV1().ServiceAccounts("default").Get(context.TODO(), saname, metav1.GetOptions{})
+		gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+		secretname := saresult.Secrets[0].Name
```
Here may panic, during my testing, I encountered once:
```log
•! Panic [0.049 seconds]
aggregatedapi testing
/root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:13
  when Serviceaccount access test
  /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:102
    Serviceaccount access test
    /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:103
      Serviceaccount have no  access in the member2 cluster [It]
      /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:112

      Test Panicked
      runtime error: index out of range [0] with length 0
      /usr/local/go/src/runtime/panic.go:90

      Full Stack Trace
      github.com/karmada-io/karmada/test/e2e/framework.WaitSecretPresentOnCluster.func1()
      	/root/go/src/github.com/karmada-io/karmada/test/e2e/framework/clusterrole.go:160 +0x3c9
      github.com/karmada-io/karmada/test/e2e/framework.WaitSecretPresentOnCluster({0x37b5c68, 0xc000192780}, {0x33be31c, 0x3})
      	/root/go/src/github.com/karmada-io/karmada/test/e2e/framework/clusterrole.go:157 +0x106
      github.com/karmada-io/karmada/test/e2e.glob..func1.3.1.2()
      	/root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:114 +0xdc
      github.com/karmada-io/karmada/test/e2e.TestE2E(0x0)
      	/root/go/src/github.com/karmada-io/karmada/test/e2e/suite_test.go:79 +0xe5
      testing.tRunner(0xc0001824e0, 0x34a62b8)
      	/usr/local/go/src/testing/testing.go:1259 +0x230
      created by testing.(*T).Run
      	/usr/local/go/src/testing/testing.go:1306 +0x727
```

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,227 @@
+
+func WaitSecretPresentOnCluster(client kubernetes.Interface, saname string) {
+	ginkgo.By(fmt.Sprintf("wait for secret (%s) present in cluster ", saname), func() {
+		saresult, err := client.CoreV1().ServiceAccounts("default").Get(context.TODO(), saname, metav1.GetOptions{})
```
Here we may replace `default` with serviceaccount's namespace.

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,227 @@
+	})
+}
+
+//WaitClusterrolebindingPresentOnCluster wait for Clusterrolebinding ready in cluster until timeout
```
Add a space at the beginning.

XiShanYongYe-Chang commented on test/e2e/aggregatedapi_test.go:
```diff
@@ -0,0 +1,161 @@
+				framework.WaitSecretPresentOnCluster(kubeClient,saName)
+				apiPath := "/apis/cluster.karmada.io/v1alpha1/clusters/" + member1 + "/proxy/apis"
+				code := framework.GetClusterNode(apiPath, kubeClient, karmadaClient, saName)
+				gomega.Expect(code).Should(gomega.Equal(200))
```
Errors occur occasionally here:
```log
aggregatedapi testing when Serviceaccount access test Serviceaccount access test 
  Serviceaccount  access in the member1 cluster
  /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:104
STEP: Creating serviceaccount(tom) in membercluster (member1)
STEP: Creating serviceaccount(tom) in karmadaCluster
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(cluster-proxy-clusterrolebinding)
STEP: wait for clusterRoleBinding(cluster-proxy-clusterrolebinding) present  
STEP: wait for secret (tom) present in cluster 
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(cluster-proxy-clusterrolebinding)

• Failure [0.146 seconds]
aggregatedapi testing
/root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:13
  when Serviceaccount access test
  /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:102
    Serviceaccount access test
    /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:103
      Serviceaccount  access in the member1 cluster [It]
      /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:104

      Expected
          <int>: 403
      to equal
          <int>: 200

      /root/go/src/github.com/karmada-io/karmada/test/e2e/aggregatedapi_test.go:109
```

XiShanYongYe-Chang reviewed (commented):
We can use `gofmt` to format the files.

XiShanYongYe-Chang commented:
Hi @wwwnay, are you still working with this e2e test?

wwwnay commented:
> Hi @wwwnay, are you still working with this e2e test?

ah.sorry, I didn't realized I was pushed failed. 

```
I0321 18:42:53.786107  125910 cluster.go:162] Got 3 member cluster and all in ready state.
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  access in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:65
STEP: Creating serviceaccount(tom) in membercluster (member1) in namespace karmadatest-7z6
STEP: Creating serviceaccount(tom) in karmadaCluster in namespace karmadatest-7z6
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(tom)
STEP: wait for serviceaccount (tom) present in karmadacluster  namespace karmadatest-7z6
STEP: wait for serviceaccount (tom) present in member(member1) cluster in namespace karmadatest-7z6
STEP: wait for clusterRoleBinding(tom) present
STEP: wait for secret (tom) present in cluster  in namespace karmadatest-7z6
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(tom)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount have no  access in the member2 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:74
STEP: Creating serviceaccount(tom) in membercluster (member1) in namespace karmadatest-7z6
STEP: Creating serviceaccount(tom) in karmadaCluster in namespace karmadatest-7z6
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(tom)
STEP: wait for serviceaccount (tom) present in karmadacluster  namespace karmadatest-7z6
STEP: wait for serviceaccount (tom) present in member(member1) cluster in namespace karmadatest-7z6
STEP: wait for clusterRoleBinding(tom) present
STEP: wait for secret (tom) present in cluster  in namespace karmadatest-7z6
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(tom)
•
------------------------------
aggregatedapi testing when Serviceaccount access test Serviceaccount access test
  Serviceaccount  does not have any permissions in the member1 cluster
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:84
STEP: Creating serviceaccount(tom) in membercluster (member1) in namespace karmadatest-7z6
STEP: Creating serviceaccount(tom) in karmadaCluster in namespace karmadatest-7z6
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(tom)
STEP: wait for serviceaccount (tom) present in karmadacluster  namespace karmadatest-7z6
STEP: wait for serviceaccount (tom) present in member(member1) cluster in namespace karmadatest-7z6
STEP: wait for clusterRoleBinding(tom) present
STEP: wait for secret (tom) present in cluster  in namespace karmadatest-7z6
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(tom)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant permission to Serviceaccount in member1 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:107
STEP: Creating serviceaccount(tom) in membercluster (member1) in namespace karmadatest-7z6
STEP: Creating serviceaccount(tom) in karmadaCluster in namespace karmadatest-7z6
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(tom)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for serviceaccount (tom) present in karmadacluster  namespace karmadatest-7z6
STEP: wait for serviceaccount (tom) present in member(member1) cluster in namespace karmadatest-7z6
STEP: wait for clusterRoleBinding(tom) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster  in namespace karmadatest-7z6
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(tom)
•
------------------------------
aggregatedapi testing when Grant permission to Serviceaccount in member
  Grant no permission to Serviceaccount in member2 cluster.
  /root/go/src/karmada/test/e2e/aggregatedapi_test.go:117
STEP: Creating serviceaccount(tom) in membercluster (member1) in namespace karmadatest-7z6
STEP: Creating serviceaccount(tom) in karmadaCluster in namespace karmadatest-7z6
STEP: Creating ClusterRole(cluster-proxy-clusterrole)
STEP: Creating ClusterRoleBinding(tom)
STEP: Creating member clusterrole(tom) in clustermember(member1)
STEP: Creating member clusterrolebinding(tom)
STEP: wait for serviceaccount (tom) present in karmadacluster  namespace karmadatest-7z6
STEP: wait for serviceaccount (tom) present in member(member1) cluster in namespace karmadatest-7z6
STEP: wait for clusterRoleBinding(tom) present
STEP: wait for clusterRoleBinding(tom) present in member(member1) cluster
STEP: wait for secret (tom) present in cluster  in namespace karmadatest-7z6
STEP: Remove clusterRole(tom) in member(member1)
STEP: Remove clusterRoleBinding(tom) in member(member1)
STEP: Remove serviceaccount(tom) in membercluster (member1)
STEP: Remove serviceaccount(tom) in karmadaCluster
STEP: Remove ClusterRole(cluster-proxy-clusterrole)
STEP: Remove ClusterRole(tom)
•
Ran 5 of 5 Specs in 3.778 seconds
SUCCESS! -- 5 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,243 @@
+	karmada "github.com/karmada-io/karmada/pkg/generated/clientset/versioned"
+)
+
+// CreateClusterrole create clusterrole.
+func Createclusterrole(client kubernetes.Interface, clusterrole *rbacv1.ClusterRole) {
```
Comment start should be same with func name, just as:
```go
// CreateClusterRole create clusterrole.
func CreateClusterRole(client kubernetes.Interface, clusterrole *rbacv1.ClusterRole)
```
Same as other parts.

XiShanYongYe-Chang commented on test/e2e/framework/clusterrole.go:
```diff
@@ -0,0 +1,243 @@
+	})
+}
+
+//  RemoveMemberclusterRole delete member cluster clusterRole.
```
There need only one space at the begin of comment:
```go
// RemoveMemberclusterRole delete member cluster clusterRole.
```
Same as other parts.

XiShanYongYe-Chang reviewed (commented):
Well done, bascically LGTM.

RainbowMango commented:
There is a conflict?

RainbowMango commented:
@wwwnay Thanks for doing this.

Would you mind I and @XiShanYongYe-Chang help with you and moving this PR forward? (I can push code to your branch)

wwwnay commented:
> @wwwnay Thanks for doing this.
> 
> Would you mind I and @XiShanYongYe-Chang help with you and moving this PR forward? (I can push code to your branch)

yes,please help me with this

RainbowMango commented:
Ok. @XiShanYongYe-Chang is working on it and after that, I'll help push the changes to your branch.
Just wait for a while.

XiShanYongYe-Chang commented:
Hi @RainbowMango @wwwnay, I refactor the related code on my fork repository:

https://github.com/XiShanYongYe-Chang/karmada/tree/pr1400

XiShanYongYe-Chang commented:
Hi @wwwnay, can you have a look?

wwwnay commented:
> Hi @wwwnay, can you have a look?

thanks, it's fine with me , much better than what i do

XiShanYongYe-Chang commented:
> thanks, it's fine with me , much better than what i do

Thanks! thank you very much too! Would you like to add more E2E or UT for karmada?

wwwnay commented:
> > thanks, it's fine with me , much better than what i do
> 
> Thanks! thank you very much too! Would you like to add more E2E or UT for karmada?

yes,i am interested， i will keep trying to add E2E or UT for karmada

RainbowMango commented:
@wwwnay, thanks again.
Here is the Github workflow guideline it might help you.

RainbowMango reviewed (approved):
/approve
/lgtm

XiShanYongYe-Chang commented:
> @wwwnay, thanks again. Here is the Github workflow guideline it might help you.

Doc site: https://github.com/karmada-io/karmada/blob/master/docs/contributors/guide/github-workflow.md

RainbowMango commented:
Sorry, I missed the link. thanks @XiShanYongYe-Chang

wwwnay commented:
> > @wwwnay, thanks again. Here is the Github workflow guideline it might help you.
> 
> Doc site: https://github.com/karmada-io/karmada/blob/master/docs/contributors/guide/github-workflow.md

@RainbowMango ,I will fllow the workflow guideline next time .
@XiShanYongYe-Chang thanks again for help me with this.

XiShanYongYe-Chang reviewed (commented):
/lgtm
/approve
